### ISSUE TITLE
Invalid TX corrupts the local node's blockchain state

### ISSUE THREAD

**Issue by EnchanterIO** [lightstreams-network/lightchain#70]
## Steps to reproduce:

1. Submit a valid TX (none 1)
1. Submit an invalid TX (e.g. too big of a size > 32kb) (nonce 2)
1. Submit a valid TX (nonce 3), this TX will fail due to a nonce mismatch in a state (nonce 2)

## Additional details

The above behaviour is due to different states being maintained across the application. In this case, the issue is TX Pool forcing a new nonce and not unregistering invalid TXs that didn't get through. 

## How to reproduce

* Run a node
* Execute the `workload` truffle test and wait for first txs to be broadcasted
```
truffle test ./test/04_workload.js --network=sirius
```
* Shut down now in the middle of the test execution
* Wait for few minutes and boot the node again.
* Every new txs signed by the same account used to performance the test will provide the following error output
```
ERROR[02-27|15:16:05.557] Nonce not strictly increasing. Expected 6926 got 6928 engine=consensus module=ABCI
```

## Solution

- Ensure an invalid TX doesn't get added to the pool
- If that happens because we have more validation logic on `ABCI::checkTx()` side than Ethereum in `TxPool::validateTx()`, remove the TX from the pool or reset the nonce.

*Note: I will first submit a PR cleaning the codebase, adding doc adding more logging etc and then a separate PR with a fix*

**Comment by EnchanterIO** [comment]
## Update: 

The above commits were attempting to bring more encapsulation to the ABCI package and while it looks like it was successful, after more thought I came to conclusion it wasn't. Writing object oriented code in GoLang while using pointers looks visually better, more comprehensive but it couples things even more because it pushes the shared state, dependency, behind more abstraction which only hides the problem under the carpet.

A different approach is needed where the ABCI dependencies and all the states are properly decoupled from the legacy codebase. In order to do so, we should first have solid tests covering the bases to ensure the changes behave as expected. I will close this task by only documenting the ABCI and refactoring naming conventions across the ABCI implementation but won't introduce more changes till tests are in place.

A new PR from branch: feature/70-abci_docs will be created in few mins.

**Comment by EnchanterIO** [comment]
Tracing is down, now blocked till all the absolutely necessary mainet tasks are finished and then we fix the sad path that corrupts local account's nonce. If this happens, here are instructions how to easily fix the problem: https://github.com/lightstreams-network/lightchain/#troubleshooting

**Comment by ggarri** [comment]
Fixed on last commit 0042c87583e9d74c57f91c40b43b6c4188e17a4c
